In [1]:
import pandas as pd
import numpy as np
data = np.array(['a','b','c','d'])
s1 = pd.Series(data)
print(s1)

0    a
1    b
2    c
3    d
dtype: object


In [2]:
data = np.array([1,2,3,4])
s2 = pd.Series(data)
print(s2)

0    1
1    2
2    3
3    4
dtype: int32


variable type in table are difference

In [3]:
data = np.array([1,'a',3,'b'])
s3 = pd.Series(data)
print(s3)

0    1
1    a
2    3
3    b
dtype: object


Type object

In [4]:
data = np.array(['a','b','c','d'])
s = pd.Series(data,index=[100,101,102,103])
print(s)

100    a
101    b
102    c
103    d
dtype: object


yes, it is change

In [5]:
s = pd.Series([1,2,3,4,5],index = ['a','b','c','d','e'])
print(s[0])

1


# Exercise: Write a code to retrieve (show) only the first three elements in Series s

In [6]:
print(s[0],s[1],s[2])

1 2 3


# Exercise: Let try to use an index (e.g. ‘a’) to access the element in Series s

In [7]:
print(s[['a','b','c','d','e']])

a    1
b    2
c    3
d    4
e    5
dtype: int64


In [8]:
import pandas as pd
df = pd.DataFrame()
print(df)

Empty DataFrame
Columns: []
Index: []


In [9]:
data = [1,2,3,4,5]
df1 = pd.DataFrame(data)
print(df1)

   0
0  1
1  2
2  3
3  4
4  5


In [10]:
data = [['Alex',10],['Bob',12],['Clarke',13]]
df2 = pd.DataFrame(data,columns=['Name','Age'])
print(df2)

     Name  Age
0    Alex   10
1     Bob   12
2  Clarke   13


In [11]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],
'Age':[28,34,29,42]}
df = pd.DataFrame(data)
print(df)

    Name  Age
0    Tom   28
1   Jack   34
2  Steve   29
3  Ricky   42


In [12]:
d = {'one' : pd.Series([1, 2, 3], index=['a', 'b', 'c']),
'two' : pd.Series([1, 2, 3, 4], index=['a', 'b', 'c', 'd'])}
df = pd.DataFrame(d)
print(df)

   one  two
a  1.0    1
b  2.0    2
c  3.0    3
d  NaN    4


we have NaN because the first series not have the 4th elemental

# Exercise: now, this is your turn to create the following table as Dataframe and name this table as studentDF.

In [13]:
data = {'Math' : pd.Series([90,10,34,23,43], index=[0,1, 2, 3,4]),
'C programming' : pd.Series([70, 20, 25, 32,73], index=[0,1, 2, 3,4]),
'Java' : pd.Series([50, 41, 67, 34,23], index=[0,1, 2, 3,4]),
'Web programming' : pd.Series([55, 66, 23, 76,67], index=[0,1, 2, 3,4])}
print("\tTable 1: studentDF Dataframe")
data = pd.DataFrame(data)
print(data)

	Table 1: studentDF Dataframe
   Math  C programming  Java  Web programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67


In [14]:
print(data)
4
print(data[2:4]) #Slice Rows
print(data.loc[0])
print(data.loc[2])
print(data.loc[0:2])

   Math  C programming  Java  Web programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
   Math  C programming  Java  Web programming
2    34             25    67               23
3    23             32    34               76
Math               90
C programming      70
Java               50
Web programming    55
Name: 0, dtype: int64
Math               34
C programming      25
Java               67
Web programming    23
Name: 2, dtype: int64
   Math  C programming  Java  Web programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23


In [15]:
df = pd.DataFrame([[1, 2], [3, 4]], columns = ['a','b'])
df2 = pd.DataFrame([[5, 6], [7, 8]], columns = ['a','b'])
df = df.append(df2)
print(df)

   a  b
0  1  2
1  3  4
0  5  6
1  7  8


# Exercise: use append to add the following table to table studentDF that you have created in the previous exercise.

In [16]:
d = {'Math' : pd.Series([59,41], index=[5,6]),
     'C programming' : pd.Series([37,34], index=[5,6]),
     'Java' : pd.Series([84,73], index=[5,6]),
     'Web programming' : pd.Series([82,55], index=[5,6])}
df = pd.DataFrame(d)
data =data.append(df)
print(data)

   Math  C programming  Java  Web programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
5    59             37    84               82
6    41             34    73               55


# Exercise: use drop function to remove the last row from table studentDF.

In [17]:
data = data.drop(6)
print(data)

   Math  C programming  Java  Web programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
5    59             37    84               82


# Milestone 1

# Exercise: try to apply sum() with table studentDF. What is the result?

In [18]:
df = pd.DataFrame(data)
print(df.sum())

Math               259
C programming      257
Java               299
Web programming    369
dtype: int64


# Exercise: try to use axis attribute to change the orientation of the summation.

In [19]:
df = pd.DataFrame(data)
print(df.sum(axis=1))

0    265
1    137
2    149
3    165
4    206
5    262
dtype: int64


# Exercise: use the above statistical functions on table studentDF, and see the results.


In [20]:
print('Count ', df.count())
print('Sum' ,df.sum())
print('Mean ', df.mean())
print('Median ', df.median)
print('Mode ', df.mode())
print('STD ', df.std())
print('Min ', df.min())
print('Max ', df.max())
print('ABS ', df.abs())
print('Prod ', df.prod())
print('Cumsum ', df.cumsum())
print('Cumprod ', df.cumprod())

Count  Math               6
C programming      6
Java               6
Web programming    6
dtype: int64
Sum Math               259
C programming      257
Java               299
Web programming    369
dtype: int64
Mean  Math               43.166667
C programming      42.833333
Java               49.833333
Web programming    61.500000
dtype: float64
Median  <bound method NDFrame._add_numeric_operations.<locals>.median of    Math  C programming  Java  Web programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
5    59             37    84               82>
Mode     Math  C programming  Java  Web programming
0    10             20    23               23
1    23             25    34               55
2    34             32    41               66
3    43             37    50               67
4    59           

# Exercise: use describe() on table studentDF and answer the following questions


In [21]:
df.describe()

Math  C programming       Java  Web programming
count   6.000000       6.000000   6.000000         6.000000
mean   43.166667      42.833333  49.833333        61.500000
std    28.407159      22.973173  22.409076        21.002381
min    10.000000      20.000000  23.000000        23.000000
25%    25.750000      26.750000  35.750000        57.750000
50%    38.500000      34.500000  45.500000        66.500000
75%    55.000000      61.750000  62.750000        73.750000
max    90.000000      73.000000  84.000000        82.000000

1.What is the mean value of Math score? = 43.166667

2.What is the score at the third quartile of Java? = 62.75000

3.What is the standard deviation of Web programming score? = 21.002381

In [22]:
df = df.reindex([0,1,2,3,4,5,6,7])
print(df)

   Math  C programming  Java  Web programming
0  90.0           70.0  50.0             55.0
1  10.0           20.0  41.0             66.0
2  34.0           25.0  67.0             23.0
3  23.0           32.0  34.0             76.0
4  43.0           73.0  23.0             67.0
5  59.0           37.0  84.0             82.0
6   NaN            NaN   NaN              NaN
7   NaN            NaN   NaN              NaN


# Exercise: use isnull() and notnull() with table studentDF and see the results.

In [23]:
print(df.isnull())
print(df.notnull())

    Math  C programming   Java  Web programming
0  False          False  False            False
1  False          False  False            False
2  False          False  False            False
3  False          False  False            False
4  False          False  False            False
5  False          False  False            False
6   True           True   True             True
7   True           True   True             True
    Math  C programming   Java  Web programming
0   True           True   True             True
1   True           True   True             True
2   True           True   True             True
3   True           True   True             True
4   True           True   True             True
5   True           True   True             True
6  False          False  False            False
7  False          False  False            False


isnull check is null or not if null return true

notnull check is null or not if null return false

In [24]:
s1 = df
s2 = df
s1 = s1.fillna(0)
s2 = s2.fillna(method='pad')
print(s1)
print(s2)

   Math  C programming  Java  Web programming
0  90.0           70.0  50.0             55.0
1  10.0           20.0  41.0             66.0
2  34.0           25.0  67.0             23.0
3  23.0           32.0  34.0             76.0
4  43.0           73.0  23.0             67.0
5  59.0           37.0  84.0             82.0
6   0.0            0.0   0.0              0.0
7   0.0            0.0   0.0              0.0
   Math  C programming  Java  Web programming
0  90.0           70.0  50.0             55.0
1  10.0           20.0  41.0             66.0
2  34.0           25.0  67.0             23.0
3  23.0           32.0  34.0             76.0
4  43.0           73.0  23.0             67.0
5  59.0           37.0  84.0             82.0
6  59.0           37.0  84.0             82.0
7  59.0           37.0  84.0             82.0


# Exercise: What is the differences between s1 and s2.

for s2 6 and 7 not 0

# Milestone 2

In [25]:
left = pd.DataFrame({
'id':[1,2,3,4,5],
'Name': ['Alex', 'Amy', 'Allen', 'Alice', 'Ayoung'],
'subject_id':['sub1','sub2','sub4','sub6','sub5']})
right = pd.DataFrame(
{'id':[1,2,3,4,5],
'subject_name':['Math','English','Social Science','Programming','Physics']})
print(left)
print(right)

   id    Name subject_id
0   1    Alex       sub1
1   2     Amy       sub2
2   3   Allen       sub4
3   4   Alice       sub6
4   5  Ayoung       sub5
   id    subject_name
0   1            Math
1   2         English
2   3  Social Science
3   4     Programming
4   5         Physics


In [26]:
merged_table = pd.merge(left,right,on='id')
merged_table

id    Name subject_id    subject_name
0   1    Alex       sub1            Math
1   2     Amy       sub2         English
2   3   Allen       sub4  Social Science
3   4   Alice       sub6     Programming
4   5  Ayoung       sub5         Physics

# Exercise: create table section and create table student

In [27]:
section = pd.DataFrame({
'id':[0,1,2,3,4,5],
'section': ['one', 'one', 'one', 'two', 'two','two']})
student = pd.DataFrame({
'id':[0,1,2,3,4,5],
'name': ['Mat', 'Mike', 'John', 'Peecha', 'Prabu','Bunchu'],
'math': [34, 12, 34, 12, 65, 23],
'eng': [23, 13, 51, 62, 12, 42] })
merged_table = pd.merge(section,student,on='id')
merged_table

id section    name  math  eng
0   0     one     Mat    34   23
1   1     one    Mike    12   13
2   2     one    John    34   51
3   3     two  Peecha    12   62
4   4     two   Prabu    65   12
5   5     two  Bunchu    23   42

In [28]:
score_sec = merged_table.groupby('section')
score_sec.sum()

id  math  eng
section               
one       3    80   87
two      12   100  116

# Milestone 3